### ISSUE TITLE
len check on tuple causes type check violation later on when accessing element 0

### ISSUE THREAD

**Issue by FatsieFS** [microsoft/pyright#10907]
## Environment data

<img width="1335" height="898" alt="Image" src="https://github.com/user-attachments/assets/4f38341c-b3af-413f-b43c-114f44b0e178" />

-   OS and version: Mint 22.2
-   Python version (& distribution if applicable, e.g. Anaconda): 3.12.3

## Code Snippet

```python
T = tuple[float, ...]

def f(t: T):
    t1 = t[1] if len(t) > 1 else 0.0
    t0 = t[0]

    return t0 + t1
```

## Repro Steps

Just entering code into python text window gives the type checking violation. If the `len(t) > 1` is removed by `t1 = 0.0` type check error goes away.

## Expected behavior

No type violation

## Actual behavior

Index 0 is out of range for type T

**Comment by erictraut** [comment]
Pyright is working as intended here, so this isn't a bug.

The `len(t)` operation applies type narrowing. For more details about type narrowing, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-narrowing).

It looks like your code assumes that the input tuple has a length of at least 1, since it's accessing `t[0]` without a guard. To make your code type safe, you can modify the input parameter type to reflect this assumption.

```python
def f(t: tuple[float, *tuple[float, ...]]): ...
```
